### PR TITLE
esp32s2/i2s: add support for I2S peripheral

### PR DESCRIPTION
**Pull request by tmedicci**
## Summary

This PR includes the I2S peripheral drivers for ESP32-S2, bindings for the generic I2S audio driver, and the CS4344 audio codec driver. Also, the I2S character driver is used for demonstration purposes.
Documentation was updated to show usage examples on ESP32-S2-Saola-1 board. 

Additionally, documentation about Espressif's boards was revisited to include hyperlinks and other references.

## Impact

No impact on existing software.

## Testing

Internal CI. Hardware testing on ESP32-S2-Saola-1 board with the CS4344 audio codec.

### CODE REVIEW

**Review by xiaoxiang781216 — APPROVED** [review]

File: arch/xtensa/src/esp32s2/esp32s2_dma.c [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+
+#include "hardware/esp32s2_dma.h"
+#include "esp32s2_dma.h"
+
+/****************************************************************************
+ * Preprocessor Definitions
+ ****************************************************************************/
+
+#ifndef MIN
+#  define MIN(a,b) (a < b ? a : b)
```
Comment (pkarashchenko) [inline arch/xtensa/src/esp32s2/esp32s2_dma.c:38]:
```suggestion
#  define MIN(a,b) ((a) < (b) ? (a) : (b))
```

File: arch/xtensa/src/esp32s2/esp32s2_i2s.c [hunk 2]
```diff
@@ -0,0 +1,1530 @@
...
+#define I2S_LL_MCLK_DIVIDER_BIT_WIDTH   (6)
+#define I2S_LL_MCLK_DIVIDER_MAX         ((1 << I2S_LL_MCLK_DIVIDER_BIT_WIDTH) - 1)
+
+/* I2S DMA channel number */
+
+#define I2S_DMA_CHANNEL_MAX (2)
+
+#ifdef CONFIG_ESP32S2_I2S_TX
+  #define I2S_TX_ENABLED 1
+  #define I2S_HAVE_TX 1
```
Comment (pkarashchenko) [inline arch/xtensa/src/esp32s2/esp32s2_i2s.c:85]:
```suggestion
#  define I2S_TX_ENABLED 1
#  define I2S_HAVE_TX 1
```

File: arch/xtensa/src/esp32s2/esp32s2_i2s.c [hunk 3]
```diff
@@ -0,0 +1,1530 @@
...
+#else
+  #define I2S_TX_ENABLED 0
```
Comment (pkarashchenko) [inline arch/xtensa/src/esp32s2/esp32s2_i2s.c:87]:
```suggestion
#  define I2S_TX_ENABLED 0
```

File: arch/xtensa/src/esp32s2/esp32s2_i2s.c [hunk 4]
```diff
@@ -0,0 +1,1530 @@
...
+#endif
+
+#ifdef CONFIG_ESP32S2_I2S_RX
+  #define I2S_RX_ENABLED 1
+#else
+  #define I2S_RX_ENABLED 0
```
Comment (pkarashchenko) [inline arch/xtensa/src/esp32s2/esp32s2_i2s.c:93]:
```suggestion
#  define I2S_RX_ENABLED 1
#else
#  define I2S_RX_ENABLED 0
```

File: arch/xtensa/src/esp32s2/esp32s2_i2s.c [hunk 5]
```diff
@@ -0,0 +1,1530 @@
...
+#ifdef CONFIG_ESP32S2_I2S_DUMPBUFFERS
+#  define       i2s_dump_buffer(m,b,s) lib_dumpbuffer(m,b,s)
+#else
+#  define       i2s_dump_buffer(m,b,s)
+#endif
+
+/* Semaphore helpers */
+
+static int      i2s_exclsem_take(struct esp32s2_i2s_s *priv);
+#define         i2s_exclsem_give(priv) nxsem_post(&priv->exclsem)
```
Comment (pkarashchenko) [inline arch/xtensa/src/esp32s2/esp32s2_i2s.c:269]: I think we should use `nxmutex`
cc: @xiaoxiang781216

File: arch/xtensa/src/esp32s2/esp32s2_i2s.c [hunk 6]
```diff
@@ -0,0 +1,1530 @@
...
+  sq_queue_t done;              /* A queue of completed transfers */
+  struct work_s work;           /* Supports worker thread operations */
+};
+
+/* The state of the one I2S peripheral */
+
+struct esp32s2_i2s_s
+{
+  struct i2s_dev_s  dev;        /* Externally visible I2S interface */
+  sem_t             exclsem;    /* Assures mutually exclusive access */
```
Comment (pkarashchenko) [inline arch/xtensa/src/esp32s2/esp32s2_i2s.c:229]: use nxmutex

File: arch/xtensa/src/esp32s2/esp32s2_dma.h [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+
+/****************************************************************************
+ * Public Function Prototypes
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: esp32s2_dma_init
+ *
+ * Description:
+ *   Initialize ESP32S2 DMA descriptions and bind the target
```
Comment (gustavonihei) [inline arch/xtensa/src/esp32s2/esp32s2_dma.h:75]:
```suggestion
 *   Initialize ESP32-S2 DMA descriptions and bind the target
```

File: arch/xtensa/src/esp32s2/esp32s2_dma.h [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+ *   buffer to these ESP32S2 DMA descriptions.
+ *
+ * Input Parameters:
+ *   dmadesc - DMA descriptions pointer
+ *   num     - DMA descriptions number
+ *   pbuf    - RX/TX buffer pointer
+ *   len     - RX/TX buffer length
```
Comment (gustavonihei) [inline arch/xtensa/src/esp32s2/esp32s2_dma.h:82]:
```suggestion
 *   Initialize ESP32-S2 DMA descriptors and bind the target
 *   buffer to these ESP32-S2 DMA descriptors.
 *
 * Input Parameters:
 *   dmadesc - Pointer to DMA descriptors
 *   num     - Number of DMA descriptors
 *   pbuf    - RX/TX buffer pointer
 *   len     - RX/TX buffer length
```

File: arch/xtensa/src/esp32s2/esp32s2_dma.h [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+ ****************************************************************************/
+
+/* DMA description type */
+
+struct esp32s2_dmadesc_s
+{
+  uint32_t ctrl;                    /* DMA control block */
+  uint8_t *pbuf;                    /* DMA TX/RX buffer address */
+  struct esp32s2_dmadesc_s *next;   /* Next DMA description address */
+};
```
Comment (gustavonihei) [inline arch/xtensa/src/esp32s2/esp32s2_dma.h:65]:
```suggestion
/* DMA descriptor type */

struct esp32s2_dmadesc_s
{
  uint32_t ctrl;                    /* DMA control block */
  uint8_t *pbuf;                    /* DMA TX/RX buffer address */
  struct esp32s2_dmadesc_s *next;   /* Next DMA descriptor address */
};
```

File: arch/xtensa/src/esp32s2/esp32s2_dma.h [hunk 10]
```diff
@@ -0,0 +1,124 @@
...
+ * Name: esp32s2_dma_init_with_padding
+ *
+ * Description:
+ *   Initialize DMA outlink descriptors and bind the target buffer to
+ *   these DMA descriptors. If len is not word-aligned, add a new descriptor
+ *   containing a 4-byte variable to make the outlink data world-aligned.
+ *
+ * Input Parameters:
+ *   dmadesc - DMA descriptions pointer
+ *   num     - DMA descriptions number
```
Comment (gustavonihei) [inline arch/xtensa/src/esp32s2/esp32s2_dma.h:102]:
```suggestion
 *   dmadesc - Pointer to DMA descriptors
 *   num     - Number of DMA descriptors
```

File: arch/xtensa/src/esp32s2/hardware/esp32s2_dma.h [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ ****************************************************************************/
+
+#ifndef __ARCH_XTENSA_SRC_ESP32S2_HARDWARE_ESP32S2_DMA_H
+#define __ARCH_XTENSA_SRC_ESP32S2_HARDWARE_ESP32S2_DMA_H
+
+/* DMA description */
```
Comment (gustavonihei) [inline arch/xtensa/src/esp32s2/hardware/esp32s2_dma.h:24]:
```suggestion
/* DMA descriptor */
```

File: arch/xtensa/src/esp32s2/esp32s2_dma.c (hunk 1 above)
Comment (tmedicci) [reply arch/xtensa/src/esp32s2/esp32s2_dma.c]: done!

File: arch/xtensa/src/esp32s2/esp32s2_i2s.c (hunk 2 above)
Comment (tmedicci) [reply arch/xtensa/src/esp32s2/esp32s2_i2s.c]: done!

File: arch/xtensa/src/esp32s2/esp32s2_i2s.c (hunk 3 above)
Comment (tmedicci) [reply arch/xtensa/src/esp32s2/esp32s2_i2s.c]: done!

File: arch/xtensa/src/esp32s2/esp32s2_i2s.c (hunk 4 above)
Comment (tmedicci) [reply arch/xtensa/src/esp32s2/esp32s2_i2s.c]: done!

File: arch/xtensa/src/esp32s2/esp32s2_i2s.c (hunk 6 above)
Comment (tmedicci) [reply arch/xtensa/src/esp32s2/esp32s2_i2s.c]: you're right!

File: arch/xtensa/src/esp32s2/esp32s2_i2s.c (hunk 5 above)
Comment (tmedicci) [reply arch/xtensa/src/esp32s2/esp32s2_i2s.c]: you're right!

File: arch/xtensa/src/esp32s2/esp32s2_i2s.c [hunk 12]
```diff
@@ -0,0 +1,1530 @@
...
+      goto errout_with_buf;
+    }
+
+  /* Add a reference to the audio buffer */
+
+  apb_reference(apb);
+
+  /* Initialize the buffer container structure */
+
+  bfcontainer->callback = (void *)callback;
```
[the remark this answers is not in the document]
Comment (tmedicci) [reply arch/xtensa/src/esp32s2/esp32s2_i2s.c]: indeed!

File: arch/xtensa/src/esp32s2/esp32s2_dma.h (hunk 8 above)
Comment (tmedicci) [reply arch/xtensa/src/esp32s2/esp32s2_dma.h]: I've just removed the `ESP32S2` to standardize the description compared to the other function. `descriptions` were changed to `descriptor` when they refer to the DMA.

File: arch/xtensa/src/esp32s2/esp32s2_dma.h (hunk 9 above)
Comment (tmedicci) [reply arch/xtensa/src/esp32s2/esp32s2_dma.h]: done!

File: arch/xtensa/src/esp32s2/hardware/esp32s2_dma.h (hunk 11 above)
Comment (tmedicci) [reply arch/xtensa/src/esp32s2/hardware/esp32s2_dma.h]: done!

**Comment by acassis** [comment]
Guys, something strange is happening, none job is marked as failed, but it was cancelled without any reason:
```
Some checks were not successful
23 successful and 3 cancelled checks 
```

**Review by xiaoxiang781216 — APPROVED** [review]

**Pull request merged**